engine/dlib/src/dlib/buffer.cpp

```diff
@@ -616,4 +644,117 @@ namespace dmBuffer
         buffer->m_ContentVersion++;
         return RESULT_OK;
     }
+
+    // Returns a pointer to the metadata item or 0 if not found.  Assumes buffer is valid.
+    Buffer::MetaData* FindMetaDataItem(Buffer* buffer, dmhash_t name_hash) {
+        dmArray<Buffer::MetaData*>& arr = buffer->m_MetaDataArray;
+        for (uint32_t  i=0; i<arr.Size(); i++) {
+            if (arr[i]->m_Name == name_hash) {
+                return arr[i];
+            }
+        }
+        return 0; // nothing found
+    }
+
+
+    Result GetMetaData(HBuffer hbuffer, dmhash_t name_hash, void** data, uint32_t* count, ValueType* type) {
+        Buffer* buffer = GetBuffer(g_BufferContext, hbuffer);
+        if (!buffer) {
+            return RESULT_BUFFER_INVALID;
+        }
+
+        Buffer::MetaData* item = FindMetaDataItem(buffer, name_hash);
+        if (!item) {
+            return RESULT_METADATA_NOT_EXIST;
+        }
+
+        *count = item->m_ValueCount;
+        *type = (ValueType)item->m_ValueType; // assumes value tpe is sanitized when setting the metadata
+        *data = item->m_Data;
+
+        return RESULT_OK;
+    }
+
+    Result SetMetaData(HBuffer hbuffer, dmhash_t name_hash, const void* data, uint32_t count, ValueType type) {
+        Buffer* buffer = GetBuffer(g_BufferContext, hbuffer);
+        if (!buffer) {
+            return RESULT_BUFFER_INVALID;
+        }
+        if (count == 0) {
+            return RESULT_METADATA_INVALID;
+        }
+
+        Buffer::MetaData* item = FindMetaDataItem(buffer, name_hash);
+        uint32_t values_block_size = GetSizeForValueType(type)*count; // do this once
+        if (item) {
+            // make sure the type and value count is right
+            if (item->m_ValueCount != count || item->m_ValueType != type) {
+                return RESULT_METADATA_INVALID; // TODO use a proper error status hsere
+            }
+            memcpy(item->m_Data, data, values_block_size);
+        } else {
+            dmArray<Buffer::MetaData*>& metadata_items = buffer->m_MetaDataArray;
+            if (metadata_items.Remaining() <= 0) {
+                metadata_items.OffsetCapacity(2); // TODO - shall we set an upper limit to the number of meta data entries ?
```

Comment:
Forgotten TODO?

Personally, I don't see a need to restrict the number of meta datas.
If it becomes a problem, we'll revisit it.